src/math.cpp

```diff
@@ -355,83 +527,144 @@ void randomseed(const CallContext& ctx) {
+    auto origin = Origin(UnaryOrigin{
+        .val = make_owning<Value>(ctx.arguments().get(0)),
+        .location = ctx.call_location(),
+        .reverse = [](const Value& new_value,
+                      const Value& old_value) -> std::optional<SourceChangeTree> {
+            Number n = std::get<Number>(new_value);
+            return old_value.force(std::atan(n.value));
+        }});
+    return math_helper(ctx, static_cast<double (*)(double)>(&std::tan), "tan").with_origin(origin);
 }
 
 auto to_integer(const CallContext& ctx) -> Value {
+    auto origin = Origin(UnaryOrigin{
+        .val = make_owning<Value>(ctx.arguments().get(0)),
+        .location = ctx.call_location(),
+        .reverse = [](const Value& new_value,
+                      const Value& old_value) -> std::optional<SourceChangeTree> {
+            if (new_value == Nil()) {
+                return std::nullopt; // if result is Nil, converting to an integer was not possible.
+                                     // so it isnt possible to revert anything.
+            } else {
+                return old_value.force(new_value);
+            }
+            // Second solution (i dont know which is the correct one)
+            // please select the right one in the review and delete the wrong one
+            // return old_value.force(new_value);
+        }});
+
     auto x = ctx.arguments().get(0);
 
     return std::visit(
-        overloaded{
-            [](Number n) -> Value {
-                if (std::modf(n.value, &n.value) == 0) {
-                    return Value(n.value);
-                } else {
-                    return Nil();
-                }
-            },
-            [](const String& s) -> Value {
-                std::regex pattern_number(R"(\s*-?\d+)");
-                std::regex pattern_hex(R"(\s*-?0[xX][\dA-Fa-f]+)");
-                std::regex pattern_exp(R"(\s*-?\d+\.?\d*[eE]\d+)");
-
-                if (std::regex_match(s.value, pattern_number) ||
-                    std::regex_match(s.value, pattern_hex) ||
-                    std::regex_match(s.value, pattern_exp)) {
-                    return Value(std::stoi(s.value, nullptr, 0));
-                } else {
-                    return Nil();
-                }
-            },
-            [](auto /*unused*/) -> Value { return Nil(); }},
-        x.raw());
+               overloaded{
+                   [](Number n) -> Value {
+                       if (std::modf(n.value, &n.value) == 0) {
+                           return Value(n.value);
+                       } else {
+                           return Nil();
+                       }
+                   },
+                   [](const String& s) -> Value {
+                       std::regex pattern_number(R"(\s*-?\d+)");
+                       std::regex pattern_hex(R"(\s*-?0[xX][\dA-Fa-f]+)");
+                       std::regex pattern_exp(R"(\s*-?\d+\.?\d*[eE]\d+)");
```

Comment:
Put the regexes in a static (const?) global variable so they don't get recompiled every time the function runs. Also it is a bit faster to put them all in out regex but you can still separate the string in different variables. I think I did the same thing for `Value::to_number`.